pyfr/backends/openmp/blasext.py

```diff
@@ -54,31 +54,44 @@ def run(self, queue):
 
         return CopyKernel()
 
-    def errest(self, x, y, z, *, norm):
-        if x.traits != y.traits != z.traits:
+    def reduction(self, *rs, method, norm, dt_val=None, dt_mat=None):
+        if not all(r.traits == rs[0].traits for r in rs):
             raise ValueError('Incompatible matrix types')
 
-        nblocks, nrow, *_, dtype = x.traits
-        ncola = x.ioshape[-2]
+        nblocks, nrow, *_, dtype = rs[0].traits
+        ncola = rs[0].ioshape[-2]
+
+        tplargs = dict(norm=norm, ncola=ncola, method=method)
+
+        if method == 'resid':
+            dt_type = 'scalar' if dt_val else 'matrix'
```

Comment:
I'd always take a dt scalar.

Reply:
What do you mean?

Reply:
Support scalar and `scalar*mat[idx]` as the two mechanisms.  So there is always a scalar (but 1 can be passed for those who don't care).  More generic and similar to what BLAS does with its scaling parameters.

Reply:
convmon is now defined in the base class, and resid kernels therefore generated before we know if there is a pi controller or not. If we want to implement this scalar factor, I'll need to move convmons back under each controller class I think. And with the necessary changes here, it will make things a little more complicated. Are we going to make use of this scalar factor when we have a matrix field for dtau?